### ISSUE TITLE
Mapping qualities in Blasr output in SAM format

### ISSUE THREAD

**Issue by snag-gh** [PacificBiosciences/blasr#247]
Hi, I am saving Blasr output in SAM/BAM format. I was wondering how to identify uniquely mapped reads? It looks like reads with multiple/secondary alignments are output with bitwise flag set to 256 and the mapping qualities could be 0,1, 2 or 3. 

I also see many alignments with the bit flag set to 0/16 and mapping qualities set to 254. I think these are the uniquely mapped reads. 

I don't see any mapping qualities in between these extremes. Do these mapping qualities have specific meanings? 

Thanks.

**Comment by mchaisso** [comment]
The secondary alignments with mapping quality of 254 represent different nonoverlapping subtrings of the original reads.  In general you can assume that any alignment with mapqv > 30 is unique.  A low mapqv indicates that there are multiple equivalent alignments of the same substring (with flag of 0 for the first alignment), or that there is a larger alignment (with likely high mapqv) that completely overlaps this smaller one.

The binary nature of the mapping qualities owes to the fact that there are few exact or nearly exact repeats with respect to the length of the read, depending on the genome.  I only see a very small fraction of reads in with mapqv between these in human alignments.

In general, mapqv is more well defined for the alignment of fixed length sequences than variable length single-molecule sequences.  For example, if a 1000 bp read has two alignments, one with 2 mismatches and all 1000bp are aligned, and another with 990 bp aligned with 10bp clipped for one reason or another, which alignment is the "true" one?  If you are looking at alignments in a highly duplicated region, there is a small chance some of them have are erroneously mapped but still have a high mapqv due to corner cases such as this.

-mark

**Comment by dbrowneup** [comment]
I'm a little skeptical about labeling reads with MAPQ > 30 as unique. I aligned some CCSs today that I generated with unanimity/ccs, then checked the FLAG/MAPQ as follows:

```
$ samtools view CCS_Alignments_v1.bam | awk '{print $2, $5}' | sort -n -k1 -k2 | uniq -c
    231 0 0
      3 0 3
      1 0 15
      1 0 20
      2 0 44
      1 0 47
   1085 0 254
    225 16 0
      3 16 3
      1 16 4
      1 16 18
      1 16 20
      1 16 25
   1168 16 254
   6812 256 0
     15 256 1
     18 256 3
      2 256 4
      1 256 5
      1 256 8
      2 256 9
      1 256 13
      1 256 14
      2 256 15
      1 256 18
      1 256 20
      1 256 24
      1 256 26
      1 256 28
      1 256 30
      1 256 36
      1 256 44
      1 256 47
   1333 256 254
   6769 272 0
      7 272 1
      3 272 2
     20 272 3
      1 272 4
      2 272 6
      2 272 10
      1 272 13
      1 272 15
      1 272 16
      2 272 19
      1 272 20
      1 272 21
      1 272 25
      1 272 26
      2 272 28
      2 272 30
      1 272 31
      1 272 35
      1 272 36
      1 272 38
      2 272 41
      2 272 44
      3 272 47
      1 272 50
      1 272 59
   1330 272 254
```

Lots of "secondary alignments" (bit 256) have MAPQ 254. But if they are secondary alignments for a read, that read would implicitly have more than one alignment and therefore not be uniquely mapped.

**Comment by mchaisso** [comment]
If there are two substrings of a read that map separately, e.g. a 1kbp read
maps from 1...500 to one location, and 501...1000 to another, the lower
scoring alignment will be output as not the primary alignment, but it will
have a mapping quality reflective of the quality for mapping that 500bp
subsequence.   Similarly, if the two substrings overlap, but share
significant portions that do not overlap, they will may high mapping
quality values although one is secondary.

If this isn't the case, it could be a bug in the computation of mapping
quality values.

Sequences over a few hundred basepairs typically either map to many
locations with the exact or nearly exact score leading to a low mapqv
 (e.g. 1,2,3), or they map so unambiguously there is effectively no 'decent
second-best' alignment that can be used to estimate the mapping quality
(e.g. one would be taking the log of 0).  It's pretty much impossible to
properly calibrate mapqvs at the upper range, so it is capped off at 254.

cheers,
-mark

**Comment by dbrowneup** [comment]
That makes a lot of sense, thanks for the clarification. However, I still think there's some odd stuff going on with MAPQ assignment. Take a couple examples from the m4 output of the same alignments:

Example 1
```
m160322_234126_00123_c100931072550000001823208905251692_s1_p0/125801/ccs/0_1976 scaffold_562 -9673 98.5937 0 0 1976 1976 0 266871 268851 348232 254
m160322_234126_00123_c100931072550000001823208905251692_s1_p0/125801/ccs/0_1976 scaffold_401 -2010 88.4906 0 1463 1976 1976 0 7109 7625 730252 0
m160322_234126_00123_c100931072550000001823208905251692_s1_p0/125801/ccs/0_1976 scaffold_25 -1244 67.1703 0 1477 1973 1976 1 43365 44092 115095 0
m160322_234126_00123_c100931072550000001823208905251692_s1_p0/125801/ccs/0_1976 scaffold_823 -836 86.076 0 1747 1976 1976 1 56522 56753 68930 0
m160322_234126_00123_c100931072550000001823208905251692_s1_p0/125801/ccs/0_1976 scaffold_392 -728 95.6522 0 1669 1830 1976 1 58759 58920 120039 0
m160322_234126_00123_c100931072550000001823208905251692_s1_p0/125801/ccs/0_1976 scaffold_149 -669 99.2647 0 1510 1646 1976 1 6215 6351 38777 0
m160322_234126_00123_c100931072550000001823208905251692_s1_p0/125801/ccs/0_1976 scaffold_201 -540 100 0 1510 1618 1976 0 2006 2114 247103 0
m160322_234126_00123_c100931072550000001823208905251692_s1_p0/125801/ccs/0_1976 scaffold_149 -529 99.0741 0 1510 1618 1976 0 32815 32923 38777 0
m160322_234126_00123_c100931072550000001823208905251692_s1_p0/125801/ccs/0_1976 scaffold_841 -430 100 0 1887 1973 1976 0 8603 8689 68960 0
```
This one makes sense... the first alignment is really long with high %ID and is thus scored MAPQ 254. The other alignments are short and piled on top of each other, all with MAPQ 0.

Example 2
```
m160322_234126_00123_c100931072550000001823208905251692_s1_p0/128231/ccs/0_1824 scaffold_1342 -8767 97.663 0 0 1824 1824 1 5432 7248 22526 254
m160322_234126_00123_c100931072550000001823208905251692_s1_p0/128231/ccs/0_1824 scaffold_1078 -8306 95.3804 0 1 1824 1824 1 225 2041 58922 0
m160322_234126_00123_c100931072550000001823208905251692_s1_p0/128231/ccs/0_1824 scaffold_350 -8232 94.9024 0 1 1824 1824 0 23958 25777 118885 0
m160322_234126_00123_c100931072550000001823208905251692_s1_p0/128231/ccs/0_1824 scaffold_366 -8224 94.9484 0 1 1824 1824 1 106948 108765 154149 0
m160322_234126_00123_c100931072550000001823208905251692_s1_p0/128231/ccs/0_1824 scaffold_1455 -8047 93.9805 0 1 1824 1824 1 4400 6217 10764 0
m160322_234126_00123_c100931072550000001823208905251692_s1_p0/128231/ccs/0_1824 scaffold_605 -8020 93.9674 0 1 1824 1824 0 16360 18176 31403 0
m160322_234126_00123_c100931072550000001823208905251692_s1_p0/128231/ccs/0_1824 scaffold_1148 -7950 93.3947 0 1 1824 1824 0 1223 3037 18971 0
m160322_234126_00123_c100931072550000001823208905251692_s1_p0/128231/ccs/0_1824 scaffold_605 -7836 92.9967 0 1 1824 1824 0 27825 29641 31403 0
m160322_234126_00123_c100931072550000001823208905251692_s1_p0/128231/ccs/0_1824 scaffold_605 -6505 93.6793 0 336 1824 1824 0 20432 21915 31403 0
m160322_234126_00123_c100931072550000001823208905251692_s1_p0/128231/ccs/0_1824 scaffold_350 -5055 95.1198 0 708 1824 1824 0 28273 29386 118885 0
```
This one is pretty dubious I think. The first alignment is given MAPQ 254, but all the subsequent alignments (with MAPQ 0) are almost identical, with very similar scores. Clearly the first alignment is not unique, but it's given a very high MAPQ.

Example 3
```
m160322_234126_00123_c100931072550000001823208905251692_s1_p0/113265/ccs/0_3616 scaffold_1 -7432 87.9676 0 9 1911 3616 0 262065 263959 320262 0
m160322_234126_00123_c100931072550000001823208905251692_s1_p0/113265/ccs/0_3616 scaffold_71 -7328 86.9196 0 9 1911 3616 0 41836 43745 85256 0
m160322_234126_00123_c100931072550000001823208905251692_s1_p0/113265/ccs/0_3616 scaffold_1148 -7266 87.6152 0 39 1911 3616 0 15728 17606 18971 0
m160322_234126_00123_c100931072550000001823208905251692_s1_p0/113265/ccs/0_3616 scaffold_1099 -7216 86.6901 0 9 1911 3616 1 49954 51858 53246 0
m160322_234126_00123_c100931072550000001823208905251692_s1_p0/113265/ccs/0_3616 scaffold_445 -6946 85.2647 0 9 1911 3616 1 43716 45637 105811 0
m160322_234126_00123_c100931072550000001823208905251692_s1_p0/113265/ccs/0_3616 scaffold_179 -6764 88.9971 0 1959 3616 3616 1 217018 218700 225961 0
m160322_234126_00123_c100931072550000001823208905251692_s1_p0/113265/ccs/0_3616 scaffold_79 -6695 91.9104 0 1924 3500 3616 1 105591 107139 109124 0
m160322_234126_00123_c100931072550000001823208905251692_s1_p0/113265/ccs/0_3616 scaffold_168 -6633 88.3027 0 1924 3616 3616 0 142291 143930 153062 254
m160322_234126_00123_c100931072550000001823208905251692_s1_p0/113265/ccs/0_3616 scaffold_256 -6203 86.4298 0 1924 3615 3616 1 4671 6233 306815 0
m160322_234126_00123_c100931072550000001823208905251692_s1_p0/113265/ccs/0_3616 scaffold_46 -2744 58.7395 0 9 3111 3616 1 164915 166997 324468 254
```
Another weird one. The top 7 scoring alignments all have MAPQ 0, but then the 8th highest scoring alignment gets MAPQ 254, as well as the 10th alignment, which has a substantially lower score than the rest.


Any thoughts on how to make sense of this?

**Comment by mchaisso** [comment]
Hi Dan,
  You bring up some really good examples.   For case 2, I would argue that
a high mapqv is pretty accurate.  It is a 2.3 % distance between the top
two alignments, so ~40 ish differences over the alignment, depending on how
it all adds up.  Typically a single SNV difference can increase mapq by 7
points.

That brings up your third example, which it's hard to say if blasr is
printing the correct mapqv for the first alignment.  There is a pretty big
difference between the top two alignments.  blasr favors longer alignments
for mapqv over shorter but possibly higher identity ones. Since the error
profile can vary across reads, sometimes there may be a subsequence that is
more accurate than others, and if that overlaps a repeat, there can be many
short, higher identity alignments detected elsewhere and but one would
ideally report the high mapqv of the longest alignment.

In the third case you show, the two low scoring alignments meet the
criteria that neither aligned portion of the read is a substring of the
other alignments, and at least the last one completely overlaps the higher
quality alignments at the top.  In my research I filter by mapqv and not
score, so I would see the last two.

I haven't looked at mapqv in a long time, but since blasr is being used on
more complex data now than when I wrote the mapqv calculation, it is
probably due for a recalibration.  While I do some active development on a
separate branch of blasr that can be integrated back to pacbio's, I
definitely can't look into this for a few months.

Hope that helps,
-mark

**Comment by dbrowneup** [comment]
Thank you for the feedback, it is definitely helpful to have your insights. I've thought about filtering my alignments by MAPQ, but my concern is that I'd be throwing away valid alignments and valuable information.

In the context of aligning PacBio reads to an assembled genome (which is what I'm doing, as opposed to aligning reads to reads), my impression is that repetitive sequences in the genome cause alignment piles on the read. I wrote a piece of software that uses a Numpy array to look at alignment coverage over the length of a read. The idea is to filter out alignments that form a big stack (>3 alignments) on a given region of the read, since they are probably from a repetitive sequence. The goal is to remove noisy, repetitive alignments and increase the signal of unique alignments.

In this case, I think if I applied a MAPQ filter, I probably wouldn't see any alignment piles on the reads and thus wouldn't be able to detect repetitive regions. I do think it would be useful if I could incorporate MAPQ or alignment score into this somehow, but I haven't come up with anything yet.
